src/materialized/src/lib.rs

```diff
@@ -320,3 +333,57 @@ impl Server {
         self.local_addr
     }
 }
+
+// Runs fetch_latest_version in a backoff loop until it succeeds once, prints a
+// warning if there is another version, then returns.
+fn check_version_loop(telemetry_url: String, cluster_id: String) {
+    let mut backoff = Duration::from_secs(5);
+    loop {
+        let latest_version = match fetch_latest_version(&telemetry_url, &cluster_id) {
+            Ok(version) => version,
+            Err(err) => {
+                backoff *= 2;
+                info!(
+                    "could not fetch latest version: {}; retrying in {:?}",
+                    err, backoff
+                );
+                thread::sleep(backoff);
+                continue;
+            }
+        };
+
+        if latest_version != version() {
+            warn!(
+                "A new version of materialized is available: {}.",
+                latest_version
+            );
+        }
+        return;
+    }
```

Comment:
There's a generic retry function in `ore::retry`. It would great if you could extend that API to take an initial backoff too, and then use it here with the async version of the reqwest client (see above). The most Rusty way to extend that API is probably to introduce some sort of `RetryBuilder` struct with builder methods to set the parameters.